Failed to init view stmt

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
#### Preparation
```SQL
create database user_center_cn;
use user_center_cn;

CREATE TABLE `dim_mongo_uid` (
  `id` bigint(20) NULL COMMENT "mongo中自增id",
  `info` varchar(65533) NULL COMMENT "游戏玩家信息  id|name"
) ENGINE=OLAP
DUPLICATE KEY(`id`, `info`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`id`) BUCKETS 2
PROPERTIES (
  "replication_num" = "3",
  "in_memory" = "false",
  "storage_format" = "DEFAULT"
);

CREATE TABLE `ods_w4_log` (
  `server_id` tinyint(4) NULL COMMENT "服务器id",
  `uid` int(11) NULL COMMENT "玩家id",
  `record_time` datetime NULL COMMENT "日志生成时间",
  `date` date NULL COMMENT "日期（游戏中4点到4点）",
  `eventname` varchar(60) NULL COMMENT "事件名",
  `env` varchar(20) NULL COMMENT "日志环境（内网测试、外网测试、cbt、正式）",
  `message` varchar(65533) NULL COMMENT "日志内容",
  INDEX env (`env`) USING BITMAP COMMENT 'env mark',
  INDEX eventname (`eventname`) USING BITMAP COMMENT 'eventname mark'
) ENGINE=OLAP
DUPLICATE KEY(`server_id`, `uid`, `record_time`)
PARTITION BY RANGE(`date`)
(PARTITION p20211114 VALUES [('0000-01-01'), ('2021-11-15')),
PARTITION p20211115 VALUES [('2021-11-15'), ('2021-11-16')),
PARTITION p20211116 VALUES [('2021-11-16'), ('2021-11-17')),
PARTITION p20220119 VALUES [('2022-01-19'), ('2022-01-20')),
PARTITION p20220120 VALUES [('2022-01-20'), ('2022-01-21')))
DISTRIBUTED BY HASH(`uid`, `eventname`) BUCKETS 32
PROPERTIES (
    "replication_num" = "2",
    "dynamic_partition.enable" = "true",
    "dynamic_partition.time_unit" = "DAY",
    "dynamic_partition.time_zone" = "Asia/Dubai",
    "dynamic_partition.start" = "-365",
    "dynamic_partition.end" = "3",
    "dynamic_partition.prefix" = "p",
    "dynamic_partition.buckets" = "32",
    "in_memory" = "false",
    "storage_format" = "DEFAULT"
); 

CREATE TABLE `ods_user_log` (
    `xid` bigint(20) NULL COMMENT "采集id",
    `ts` bigint(20) NULL COMMENT "采集时间戳",
    `table` varchar(60) NULL COMMENT "表名",
    `date` date NULL COMMENT "日期",
    `record_time` datetime NULL COMMENT "采集时间",
    `type` varchar(20) NULL COMMENT "操作类型",
    `commit` boolean NULL COMMENT "操作类型",
    `database` varchar(60) NULL COMMENT "数据库",
    `message` varchar(65533) NULL COMMENT "信息内容"
) ENGINE=OLAP
DUPLICATE KEY(`xid`, `ts`, `table`)
PARTITION BY RANGE(`date`)
(PARTITION p20211212 VALUES [('2021-12-12'), ('2021-12-13')),
PARTITION p20211213 VALUES [('2021-12-13'), ('2021-12-14')),
PARTITION p20220119 VALUES [('2022-01-19'), ('2022-01-20')),
PARTITION p20220120 VALUES [('2022-01-20'), ('2022-01-21')))
DISTRIBUTED BY HASH(`table`) BUCKETS 4
PROPERTIES (
    "replication_num" = "2",
    "dynamic_partition.enable" = "true",
    "dynamic_partition.time_unit" = "DAY",
    "dynamic_partition.time_zone" = "Asia/Shanghai",
    "dynamic_partition.start" = "-1000",
    "dynamic_partition.end" = "3",
    "dynamic_partition.prefix" = "p",
    "dynamic_partition.buckets" = "4",
    "in_memory" = "false",
    "storage_format" = "DEFAULT"
); 


CREATE TABLE `dim_mongo_sdk` (
  `account` varchar(65533) NULL COMMENT "账号id 长码",
  `id` bigint(20) NULL COMMENT "mongo中自增id"
) ENGINE=OLAP
DUPLICATE KEY(`account`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`account`) BUCKETS 2
PROPERTIES (
    "replication_num" = "3",
    "in_memory" = "false",
    "storage_format" = "DEFAULT"
);  

CREATE VIEW `dwd_login_view` AS SELECT to_date(from_unixtime(get_json_int(`message`, '$.login_at'))) AS `date`, get_json_int(`message`, '$.channel_id') AS `channel_id`, from_unixtime(get_json_int(`message`, '$.login_at')) AS `login_at`, from_unixtime(get_json_int(`message`, '$.created_at')) AS `created_at`, get_json_int(`message`, '$.login_log_id') AS `login_log_id`, get_json_int(`message`, '$.account_id') AS `account_id`, get_json_string(`message`, '$.long_account_id') AS `long_account_id`, get_json_string(`message`, '$.short_account_id') AS `short_account_id`, get_json_string(`message`, '$.game_code') AS `game_code`, get_json_string(`message`, '$.package_code') AS `package_code`, get_json_string(`message`, '$.login_ip') AS `login_ip`, get_json_string(`message`, '$.device_ua') AS `device_ua`, get_json_string(`message`, '$.device_id') AS `device_id`, get_json_string(`message`, '$.device_net') AS `device_net`, get_json_string(`message`, '$.device_ratio') AS `device_ratio`, get_json_string(`message`, '$.device_os') AS `device_os`, get_json_string(`message`, '$.device_carrier') AS `device_carrier`, get_json_string(`message`, '$.ext') AS `ext` FROM `default_cluster:user_center_cn`.`ods_user_log` WHERE `table` LIKE 'login_log%';

CREATE VIEW `dim_login_view` AS SELECT `user_center_cn`.`dwd_login_view`.`channel_id` AS `channel_id`, `user_center_cn`.`dwd_login_view`.`login_at` AS `login_at`, `user_center_cn`.`dwd_login_view`.`date` AS `date`, `user_center_cn`.`dwd_login_view`.`login_log_id` AS `login_log_id`, `user_center_cn`.`dwd_login_view`.`account_id` AS `account_id`, `user_center_cn`.`dwd_login_view`.`long_account_id` AS `long_account_id`, `user_center_cn`.`dwd_login_view`.`short_account_id` AS `short_account_id`, `user_center_cn`.`dwd_login_view`.`game_code` AS `game_code`, `user_center_cn`.`dwd_login_view`.`package_code` AS `package_code`, `user_center_cn`.`dwd_login_view`.`login_ip` AS `login_ip`, `user_center_cn`.`dwd_login_view`.`device_ua` AS `device_ua`, `user_center_cn`.`dwd_login_view`.`device_id` AS `device_id`, `user_center_cn`.`dwd_login_view`.`device_net` AS `device_net`, `user_center_cn`.`dwd_login_view`.`device_ratio` AS `device_ratio`, `user_center_cn`.`dwd_login_view`.`device_os` AS `device_os`, `user_center_cn`.`dwd_login_view`.`device_carrier`AS `device_carrier`, `user_center_cn`.`dwd_login_view`.`ext` AS `ext`, `user_center_cn`.`dwd_login_view`.`created_at` AS `created_at` FROM `user_center_cn`.`dwd_login_view` WHERE `game_code` = 'w4';

CREATE VIEW `dim_phone_view` AS SELECT to_date(from_unixtime(get_json_int(`message`, '$.created_at'))) AS `date`, from_unixtime(get_json_int(`message`, '$.created_at')) AS `record_time`, get_json_int(`message`, '$.account_id') AS `account_id`, get_json_int(`message`, '$.channel_id') AS `channel_id`, get_json_string(`message`, '$.account_name') AS `account_name`, get_json_string(`message`, '$.email') AS `email`, get_json_string(`message`, '$.identity') AS `identity`, get_json_string(`message`, '$.credential') AS `credential`, get_json_string(`message`, '$.created_at') AS `created_at` FROM `default_cluster:user_center_cn`.`ods_user_log` WHERE (`table` LIKE 'account_credential%') AND (get_json_int(`message`, '$.channel_id') = 11);

CREATE VIEW `dim_account_view` AS SELECT to_date(get_json_string(`message`, '$.created_at')) AS `date`, get_json_int(`message`, '$.account_id') AS `account_id`, get_json_string(`message`, '$.created_at') AS `created_at`, get_json_string(`message`, '$.updated_at') AS `updated_at`, get_json_string(`message`, '$.project') AS `project`, get_json_string(`message`, '$.id_num') AS `id_num`, get_json_string(`message`, '$.name') AS `name`, get_json_string(`message`, '$.pi') AS `pi`, get_json_string(`message`, '$.ai') AS `ai`, get_json_string(`message`, '$.app_id') AS `app_id`, get_json_string(`message`, '$.biz_id') AS `biz_id`, get_json_string(`message`, '$.status') AS `status`, get_json_string(`message`, '$.msg') AS `msg` FROM `default_cluster:user_center_cn`.`ods_user_log` WHERE (`table` = 'idcard') AND (get_json_string(`message`, '$.project') = 'w4');

```

#### Create View
```SQL
CREATE VIEW yao_test
AS
WITH mongo_uid AS (
		SELECT id, regexp_extract(info, '^.+id=([0-9]+).+', 1) AS UID
		FROM dim_mongo_uid
	), 
	mongo_all AS (
		SELECT u.id, UID, account
		FROM mongo_uid u
			JOIN dim_mongo_sdk s ON u.id = s.id
	), 
	dim_login AS (
		SELECT DISTINCT account_id, long_account_id
		FROM dim_login_view
	), 
	dim_phone AS (
		SELECT account_id, IDENTITY AS phone
		FROM dim_phone_view
	), 
	dim_account AS (
		SELECT account_id
			, year(curdate()) - substr(id_num, 7, 4) AS age
			, name
		FROM dim_account_view
	), 
	dim_all AS (
		SELECT l.account_id, long_account_id, name, phone, age
		FROM dim_login l
			JOIN dim_phone p ON l.account_id = p.account_id
			JOIN dim_account a ON p.account_id = a.account_id
	)
SELECT account_id, long_account_id, name, age, phone , UID
FROM dim_all d
    JOIN mongo_all m ON d.long_account_id = m.account;
```

### Expected behavior (Required)
Create view successfully.

### Real behavior (Required)
MySQL client throws the following error:
```
ERROR 1064 (HY000): failed to init view stmt
```

fe.warn.log contains the following content:
```
com.starrocks.common.DdlException: failed to init view stmt
        at com.starrocks.catalog.Catalog.createView(Catalog.java:6069) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.DdlExecutor.execute(DdlExecutor.java:194) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.StmtExecutor.handleDdlStmt(StmtExecutor.java:900) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.StmtExecutor.execute(StmtExecutor.java:420) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:268) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:415) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:651) ~[starrocks-fe.jar:?]
        at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55) ~[starrocks-fe.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_202]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_202]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_202]
Caused by: com.starrocks.common.UserException: Failed to parse view-definition statement of view: yao_test
        at com.starrocks.catalog.View.init(View.java:204) ~[starrocks-fe.jar:?]
        at com.starrocks.catalog.Catalog.createView(Catalog.java:6067) ~[starrocks-fe.jar:?]
        ... 10 more
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
b82a996d22a564e6dfc036c852d75e32b47454de

--- a/fe/fe-core/src/main/java/com/starrocks/analysis/QueryStmt.java
+++ b/fe/fe-core/src/main/java/com/starrocks/analysis/QueryStmt.java
@@ -25,6 +25,7 @@ import com.google.common.base.Preconditions;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.starrocks.catalog.Database;
+import com.starrocks.catalog.View;
 import com.starrocks.common.AnalysisException;
 import com.starrocks.common.ErrorCode;
 import com.starrocks.common.ErrorReport;
@@ -167,6 +168,9 @@ public abstract class QueryStmt extends StatementBase {
         super.analyze(analyzer);
         analyzeLimit(analyzer);
         if (hasWithClause()) {
+            for (View v : withClause_.getViews()) {
+                v.getQueryStmt().setNeedToSql(this.needToSql);
+            }
             withClause_.analyze(analyzer);
         }
         analyzeOutfile();

--- a/fe/fe-core/src/test/java/com/starrocks/sql/plan/ViewPlanTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/sql/plan/ViewPlanTest.java
@@ -1650,4 +1650,18 @@ public class ViewPlanTest extends PlanTestBase {
 
         starRocksAssert.dropView("test_view16");
     }
+
+    @Test
+    public void testWith() throws Exception {
+        String sql = "With x0 AS (SELECT DISTINCT v1, v2 FROM t0)\n" +
+                " SELECT v1, v2 from x0";
+        testView(sql);
+    }
+
+    @Test
+    public void testWithMore() throws Exception {
+        String sql = "With x0 AS (SELECT DISTINCT v1, v2 FROM t0)\n" +
+                " SELECT v1, v2 from x0";
+        testView(sql);
+    }
 }

